### PR TITLE
Formalizing the Schema

### PR DESCRIPTION
**Pull request by dgasmith**
## Description
To kick this off again, I have started creating a [json-schema](http://json-schema.org) implementation based the discussed schema along with documentation via Sphinx and RTD. As a forward I should note that the implementation here is an attempt to build something concrete. Many things can be changed, but some sort of core needs to be formed to move forward.

This has three basic parts:
 - A small python module to help organize the schema and collection previous versions.
 - The start of a set of examples that will also be used for testing.
 - A RTD-based website that is auto generated from the current repo and started porting the current markdown files to this website. This will be up once this PR is merged.

The presented schema aims to provide a very basic interface with a few notable points such as lack of unit support. Before continuing to build out this schema I wanted to get intermediate feedback about the current setup and general organization.

## Questions
- [x] Is there a good way to produce html-based documentation of schema automatically? I found several python/sphinx solution, but they do not seem to work. I also like swagger.io for REST interfaces; however, it doesn't seem to match up to the requirements of this schema.

## Status
- [x] Ready to go

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -99,3 +99,11 @@ ENV/
 
 # mypy
 .mypy_cache/
+
+# Vim leftover
```
Comment (langner) [inline .gitignore]: Not everyone uses vim, and this is not strictly related to the schema. Maybe put this in your private ignore list? (.git/info/exclude)

File: qc_schema/dev/molecule.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        },
+        "geometry": {
+            "description": "The 3N XYZ coordinates of the atoms involved.",
+            "type": "array",
+            "items": {
+                "type": "number"
+            }
+        },
+        "masses": {
+            "description": "The mass of the molecule, canonical weights assumed if not given.",
```
Comment (langner) [inline qc_schema/dev/molecule.py]: The name is plural but the description has singular mass.

File: qc_schema/dev/properties/mp_properties.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+mp_properties["mp2_double_energy"] = {
+    "type":
+    "number",
+    "description":
+    "The doubles portion of the MP2 correlation energy including same-spin and opposite-spin correlations."
+}
+
+mp_properties['mp2_total_correlation_energy'] = {
```
Comment (langner) [inline qc_schema/dev/properties/mp_properties.py]: Would be useful to also have a total MP energy property. I know it's easy to sum up, but some people like to see the total energy.

File: .gitignore (hunk 1 above)
Comment (dgasmith) [reply .gitignore]: Sure, can do. I normally see these left in the `.gitignore` files, is there a reason to move them out?

File: qc_schema/dev/properties/scf_properties.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
+"""
+The complete list of SCF level properties.
+"""
+
+scf_properties = {}
+
+scf_properties["scf_one_electron_energy"] = {
+    "description": "The one-electron energy contribution [H] to the total SCF energy.",
```
Comment (berquist) [inline qc_schema/dev/properties/scf_properties.py]: Here and later on, what is `[H]`?

Comment (dgasmith) [reply qc_schema/dev/properties/scf_properties.py]: The core Hamiltonian, I dont really look at these descriptions as complete (they are not intended to be), but more about the general scaffold to build on.

File: qc_schema/dev/properties/mp_properties.py (hunk 3 above)
Comment (dgasmith) [reply qc_schema/dev/properties/mp_properties.py]: Sure, missed this.

File: .gitignore (hunk 1 above)
Comment (langner) [reply .gitignore]: IMO projects should have just entries relevant to their code. But of course it's not super important.

File: qc_schema/dev/molecule.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            "description": "Whether to rotate the molecule to a standard orientation or not.",
+            "type": "boolean",
+            "default": False
+        },
+        "provenance": {
+            "type": "object",
+            "$ref": "#/definitions/provenance"
+        }
+    },
+    "required": ["symbols", "geometry"],
```
Comment (langner) [inline qc_schema/dev/molecule.py]: Is this actually enforced somehow?

Comment (dgasmith) [reply qc_schema/dev/molecule.py]: If you run this through a json-schema validator you will get a raise like the following:
```
E           jsonschema.exceptions.ValidationError: 'symbols' is a required property
E           Failed validating 'required' in schema['properties']['molecule']:
```

Im not sure how to require this at other levels.

File: qc_schema/dev/molecule.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    "properties": {
+        "symbols": {
+            "description": "The atom symbol for each atom in the molecule.",
+            "type": "array",
+            "items": {
+                "type": "string"
+            }
+        },
+        "geometry": {
+            "description": "The 3N XYZ coordinates of the atoms involved.",
```
Comment (loriab) [inline qc_schema/dev/molecule.py]: explicitly mention flat (3N, ) rather than nester (N, 3), I think.

File: qc_schema/dev/properties/scf_properties.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
+"""
+The complete list of SCF level properties.
+"""
+
+scf_properties = {}
+
+scf_properties["scf_one_electron_energy"] = {
+    "description": "The one-electron energy contribution to the total SCF energy.",
+    "type": "number"
+}
```
Comment (loriab) [inline qc_schema/dev/properties/scf_properties.py]: If this is meant to refer to Hartree--Fock, not HF-or-DFT, I suggest we be explicit now and use "HF".

Comment (dgasmith) [reply qc_schema/dev/properties/scf_properties.py]: What if we define that is as simply `\sum{D * H}`? I worry that splitting HF and SCF into two parts could get much more complex.

File: Topology/README.md [hunk 8]
```diff
@@ -23,15 +23,15 @@ The following are optional fields and default values (option, more a list of pos
   - `comment` (str) - Any additional comment one would attach to the molecule.
```
Comment (ghutchis) [inline Topology/README.md]: I think, based on #35 that it makes sense to have an "identifier" section, with comment, molecule name, formula, InChI, SMILES, etc. as optional.  This would probably also include provenance and DOI.

Looking at my notes, I think these were grouped in the discussion.

File: docs/source/index.rst [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+ - `VTK <http://www.vtk.org>`_
+
+**Quantum Chemistry Engines:**
+ - `MPQC <https://github.com/ValeevGroup/mpqc>`_
+ - `NWChem <http://www.nwchem-sw.org/index.php/Main_Page>`_
+ - `Psi4 <https://github.com/psi4/psi4>`_
+
+**Translators:**
+ - `cclib <http://cclib.github.io>`_
+
```
Comment (ghutchis) [inline docs/source/index.rst]: - `openbabel <http://openbabel.org/>`_

File: docs/source/index.rst [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+A concrete list of requirements for this schema can be found [here](Requirements.md).
+
+**Organizations:**
+ * `The Molecular Sciences Software Institute <http://www.molssi.org>`_
+
+**Visualizers:**
+ - `Avogadro <https://avogadro.cc>`_
+ - `Molecular Design Toolkit <https://github.com/Autodesk/molecular-design-toolkit>`_
+ - `VTK <http://www.vtk.org>`_
+
```
Comment (ghutchis) [inline docs/source/index.rst]: - `Jmol / JSmol <http://jmol.org/>`_

File: docs/source/tech_specs.rst [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+    // 1) Storing fields as tables: creates an mmCIF/PDB-like layout
+    {atoms={type:'table[atom]',
+            fields=['name', 'atomic_number', 'mass/Dalton', 'residue_index', 'position/angstrom', 'momentum/angstrom*amu*fs^-1']
+            entries=[
+              ['CA', 6, 12.0, 0, [0.214,12.124,1.12], [0,0,0]],
+              ['N', 7, 14.20, 0, [0.214,12.124,1.12], [0,0,0]],
+              ...}
+    
+    // 2) Storing fields as arrays: much more compact, but harder to read and edit
```
Comment (ghutchis) [inline docs/source/tech_specs.rst]: Perhaps its harder to read/edit as a human, but not in a program. Please don't use this language. As @loriab mentioned, this was the choice at the workshop and there are a lot of benefits on the "writing code" side.

File: docs/source/tech_specs.rst [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+ 1) As a table of values (these values do very well in HDF5 as well)
+ 2) As a set of arrays
+ 3) As a list of objects (works well in HDF5 too, would be worth brining in a HDF5 expert)
+
+Examples:
+
+.. code:: python
+
+    // 1) Storing fields as tables: creates an mmCIF/PDB-like layout
```
Comment (ghutchis) [inline docs/source/tech_specs.rst]: Please no. Let's not even suggest something *remotely* like the mess of PDB files.

File: qc_schema/dev/molecule.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+            "type": "array",
+            "items": {
+                "type": "number"
+            }
+        },
+        "name": {
+            "description": "The name of the molecule.",
+            "type": "string"
+        },
+        "comment": {
```
Comment (ghutchis) [inline qc_schema/dev/molecule.py]: Let's make these part of an object for "identifiers" to reflect feedback from #35

File: qc_schema/dev/molecule.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+        "fragment_multiplicities": {
+            "description": "A list of multiplicites associated with each fragment.",
+            "type": "array",
+            "items": {
+                "type": "number",
+                "multipleOf": 1.0
+            }
+        },
+        "fix_com": {
+            "description": "Whether to adjust to the molecule to the COM or not.",
```
Comment (ghutchis) [inline qc_schema/dev/molecule.py]: COM? I assume you mean "center of mass" but let's be explicit.

File: qc_schema/dev/properties/scf_properties.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    "type":
+    "number",
+    "description":
+    """
+The dispersion correction appended to an underlying functional
+when a DFT-D method is requested.`
+"""
+}
+
+scf_properties["scf_dipole_moment"] = {
```
Comment (ghutchis) [inline qc_schema/dev/properties/scf_properties.py]: Does this have to be 'scf' dipole moment? Maybe I'm pedantic, but what if I run CCSD(T) - most programs don't print out two different dipole moments IIRC.

File: docs/source/tech_specs.rst (hunk 11 above)
Comment (dgasmith) [reply docs/source/tech_specs.rst]: I was mostly porting over current Markdown documentation to RST as an example, but there seems to be bringing up a bit of old discussion that is indeed solved. Should we covert the solved topics over to a simpler discussion of why the choice was made?

File: Topology/README.md (hunk 8 above)
Comment (dgasmith) [reply Topology/README.md]: Are you thinking of `comment` as a dictionary? I was considering moving up a few of those to top level optional fields.

**Comment by dgasmith** [comment]
For the magic header, poking around there doesn't seem to be a full standard. However, finding a few examples this following seems popular:

```
{
  "name": "Schema name",
  "description": "A quick description",
  "url": "http://schema_host.org/schemas/v0.1/something.schema",
  "version": "v1.0",
}
```

Should this be spun off into its own issue however?

File: Topology/README.md (hunk 8 above)
Comment (ghutchis) [reply Topology/README.md]: No, I'm thinking that there's an obvious request for identifiers, and I think comment/title have usually been identifiers in QC codes.

So my suggestion is that there's an explicit identifier object - something like:

```
"identifier" : {
   "name": "aspirin",
   "comment": "training set",
   "formula": "C9H8O4",
   "smiles": "O=C(C)Oc1ccccc1C(=O)O",
   "inchikey": "BSYNRYMUTXBXSQ-UHFFFAOYSA-N"
}
```

File: qc_schema/dev/properties/scf_properties.py (hunk 15 above)
Comment (dgasmith) [reply qc_schema/dev/properties/scf_properties.py]: I think its good to clarify if someone does say CCSD bond orders. It can become unclear if you stored the SCF or CCSD dipoles.

File: docs/source/tech_specs.rst (hunk 11 above)
Comment (ghutchis) [reply docs/source/tech_specs.rst]: Yes, I think that's a good idea. As people implement, I don't think there should be ambiguity and I think a discussion of the choice is good.

(Incidentally, I'd suggest when the schema is done, that this be published in J. Cheminf.)

**Comment by ghutchis** [comment]
I'd probably go with name/version/url/description in that order, but implementations will go with alphabetical order.

Can we drop the description part? Then you'd have:

```
{
  "name": "QC_JSON",
  "url": "http://schema_host.org/schemas/v0.1/something.schema",
  "version": "v1.0",
}
```

That seems suitably unique for file/magic sensing.

**Comment by ghutchis** [comment]
IMHO, this is a great start and I'm in favor of merging this and starting new issues / pull requests.

File: qc_schema/dev/properties/scf_properties.py (hunk 15 above)
Comment (langner) [reply qc_schema/dev/properties/scf_properties.py]: Won't that lead to a proliferation of new objects? Maybe there should be a `properties` object:
```
property: {
  type: "dipole moment"
  method: "SCF"
  value: ...
}
```

**Comment by dgasmith** [comment]
Im in favor of merging as well. Certainly a lot left to do, but we can break the current topics off into issues.

File: qc_schema/dev/properties/scf_properties.py (hunk 15 above)
Comment (dgasmith) [reply qc_schema/dev/properties/scf_properties.py]: It doesn't seem like too many more fields as we will have energy definitions for these slots anyways. I might be missing your point on the object side of things however.

**Comment by cryos** [comment]
This is looking great, and would agree that merging, and then breaking off into smaller PRs/discussions would be beneficial.

**Comment by dgasmith** [comment]
Going to pull this in and setup the docs. Please spin off issues or PR's as needed to fix current issues.
